add support for defining macros on command line

josuah opened the pull request:
Command-line defined macros are useful to define specific parameters to
one invocation of a program.

For example: ls | xargs -I {} mdsort -D box={} -f ~/.mdsort/archive.conf

The free function had to be moved to util.c as it gets used by mdsort.c.

The current patch does not allow overriding macros from the config file yet.

josuah committed:
add support for defining macros on command line

Command-line defined macros are useful to define specific parameters to
one invocation of a program.

For example: ls | xargs -I {} mdsort -D box={} -f ~/.mdsort/archive.conf

mptre commented on extern.h:
```diff
@@ -411,7 +412,7 @@ struct config_list {
 };
 
-struct config_list *config_parse(const char *, const struct environment *);
+struct config_list *config_parse(const char *, const struct environment *, struct macro_list *);
```
`struct environment` is by convention the last argument

josuah replied:
I am glad I learn this convention, I always struggle to choose argument order.

mptre commented on mdsort.c:
```diff
@@ -42,8 +42,10 @@ main(int argc, char *argv[])
+	struct macro_list macros;
 	struct maildir *md;
 	struct message *msg;
+	char *eq;
```
Reduce the scope by introducing a block:

```c
case 'D': {
  char *s;
}
```

josuah replied:
all right, done

mptre commented on mdsort.c:
```diff
@@ -70,6 +74,12 @@ main(int argc, char *argv[])
+			if ((eq = strchr(optarg, '=')) == NULL)
+				err(1, "invalid macro: %s", optarg);
+			*eq = '\0';
+			macros_insertc(&macros, optarg, eq + 1);
```
`macros_insertc` must only be used when the number of macros are known at compile time, see the comment attached to the implementation:

```
$ mdsort -D a=a -D b=b -D c=c        
mdsort: macros_insertc: stack storage exhausted
```

This routine allows macros to be used without requiring any heap allocation. Instead, `macros_insert` must be used here.

mptre reviewed (commented):
Hi,
Thanks for the diff. Just to make I've understand your use case,
something along the lines is what you want to achieve?

```
$ cat ~/.mdsort/archive.conf
maildir "${box}" {
  match all move "~/Maildir/Archive"
}
```

> The current patch does not allow overriding macros from the config file yet.

I would want this to behave similar to how make(1) behaves: any macro
supplied on the command line takes higher precedence than the ones
defined in the configuration. You should be able to achieve such
semantics by adding a flags argument to `macros_insert()` and add a new
flag `MACRO_FLAG_IMMUTABLE`. Also, command line arguments must be
flagged as `MACRO_FLAG_CONST`.

josuah commented:
Thank you for the feedback.  I am interested in providing the code changes, hopefully soon...

Regarding the use-case, the exact invocation is below, to archive old mail from big mailing lists per year:

```
maildir "${box}" {
        match header "Date" / ([0-9]{4}) / and date > 6 month \
                move "${box}/\1"
}
```

A `"~/glob-*"` or a `{ "~/list" "~/of", "~/box" }` could also work, but the command-line macro feature is useful anyway.

josuah commented:
> I would want this to behave similar to how make(1) behaves: any macro
> supplied on the command line takes higher precedence than the ones
> defined in the configuration.

Here we go:

```
$ cat mdsort.conf
inbox = "~/Maildir"
stdin {
        match all move "${inbox}"
}
$ ./mdsort -f mdsort.conf -D inbox=/dev/null -
mdsort: opendir: /dev/null/new: Not a directory
```

On the other hand, there is now a new minor bug:

```
$ cat mdsort.conf
inbox = "~/Maildir"
inbox = "~/Maildir"
stdin {
        match all move "${inbox}"
}
$ ./mdsort -f mdsort.conf -D inbox=/dev/null -
mdsort: opendir: /dev/null/new: Not a directory
```

It should warn us that there is a macro redefinition there.

This would mean we have to adapt further, such as keeping all macros
definitions and have an extra parameter to macros_find to set a flag of
what kind of macro to seek.

Is it worth the cost?

> You should be able to achieve such semantics by adding a flags
> argument to macros_insert() and add a new flag MACRO_FLAG_IMMUTABLE.
> Also, command line arguments must be flagged as MACRO_FLAG_CONST.

This was implemented by having `macros_inserti()` calling `macros_insert()`
for more shared code, then appending both flags after-hand.

Should this be refactored, such as having macros_insert having an argument
to choose which flag to use?

mptre commented:
On Sun, May 02, 2021 at 02:06:23PM -0700, josuah wrote:
> > I would want this to behave similar to how make(1) behaves: any macro
> > supplied on the command line takes higher precedence than the ones
> > defined in the configuration.
> 
> Here we go:
> 
> ```
> $ cat mdsort.conf
> inbox = "~/Maildir"
> stdin {
>         match all move "${inbox}"
> }
> $ ./mdsort -f mdsort.conf -D inbox=/dev/null -
> mdsort: opendir: /dev/null/new: Not a directory

Looks good.

> On the other hand, there is now a new minor bug:
> 
> ```
> $ cat mdsort.conf
> inbox = "~/Maildir"
> inbox = "~/Maildir"
> stdin {
>         match all move "${inbox}"
> }
> $ ./mdsort -f mdsort.conf -D inbox=/dev/null -
> mdsort: opendir: /dev/null/new: Not a directory
> ```
> 
> It should warn us that there is a macro redefinition there.
> 
> This would mean we have to adapt further, such as keeping all macros
> definitions and have an extra parameter to macros_find to set a flag of
> what kind of macro to seek.
> 
> Is it worth the cost?

I think you should be able to introduce a mc_defs field to struct macro
and get the error reporting right:

	if ((mc = macros_find(macros, name)) != NULL) {
		if (mc->mc_flags & MACRO_FLAG_IMMUTABLE) {
			if (++mc->mc_defs > 1)
				return 1;
			return 0; 
		}
		return 1;
	}

> > You should be able to achieve such semantics by adding a flags
> > argument to macros_insert() and add a new flag MACRO_FLAG_IMMUTABLE.
> > Also, command line arguments must be flagged as MACRO_FLAG_CONST.
> 
> This was implemented by having `macros_inserti()` calling `macros_insert()`
> for more shared code, then appending both flags after-hand.
> 
> Should this be refactored, such as having macros_insert having an argument
> to choose which flag to use?

I would still prefer adding a flags argument to macros_insert(). This
should be a separate commit before adding support for command line
macros.

	int
	macros_insert(struct macro_list *macros, char *name, char *value,
	    unsigned int flags, int lno)
	{
	}

mptre commented:
Any updates here or have this work gone stale?

josuah committed:
add a flag arguemnt to macros_insert() to initialize the extra field

josuah commented:
These two are ready for the merge as I understood, and then I need to add the better check for existing macros in macros_insertc.

According to my measurements, "hopefully soon" means 5 months in my language.

josuah committed:
report macros with multiple definitions

The previous commit, this also permits to remove macros_insertc(),
that get merged with macros_insert().

josuah commented:
And here we go. The invocation is a compacted version from you above.

Testing with and without `-D inbox=/dev/null` gives the same result:

- error when two inbox= macros defined in the config file
- continue when one inbox= macro defined in the config file

Thank you for the guidance and patience, I let you review but no hurry.

mptre commented:
Thanks, committed with some tweaks. Take a look and see if we agree upon
the semantics.

josuah commented:
Yes, this way no need to define `mc_defs` to 1 or 0 depending on the flags, looks cleaner.
Thank you for the changes, and tests!
